src/storage/mod.rs

```diff
@@ -0,0 +1,566 @@
+///
+/// This hook returns a Signal that can be used to read and modify the state.
+/// The changes to the state will be persisted to storage and all other app sessions will be notified of the change to update their local state.
+pub fn use_synced_storage<S, T>(cx: &ScopeState, key: S::Key, init: impl FnOnce() -> T) -> Signal<T>
+where
+    S: StorageBacking + StorageSubscriber<S>,
+    T: Serialize + DeserializeOwned + Clone + Send + Sync + PartialEq + 'static,
+    S::Key: Clone,
+{
+    *cx.use_hook(|| new_synced_storage::<S, T>(cx, key, init))
+}
+
+/// Create a signal that can be used to store data that will persist across application reloads and be synced across all app sessions for a given installation or browser.
+///
+/// This hook returns a Signal that can be used to read and modify the state.
+/// The changes to the state will be persisted to storage and all other app sessions will be notified of the change to update their local state.
+pub fn new_synced_storage<S, T>(cx: &ScopeState, key: S::Key, init: impl FnOnce() -> T) -> Signal<T>
+where
+    S: StorageBacking + StorageSubscriber<S>,
+    T: Serialize + DeserializeOwned + Clone + Send + Sync + PartialEq + 'static,
+    S::Key: Clone,
+{
+    let mut init = Some(init);
+    let signal = {
+        if cfg!(feature = "ssr") {
+            // SSR does not support synced storage on the backend. We will just use a normal Signal to represent the initial state.
+            // The client will hydrate this with a correct SyncedStorageEntry and maintain state.
+            Signal::new(init.take().unwrap()())
+        } else if cfg!(feature = "hydrate") {
+            let key_clone = key.clone();
+            let storage_entry = new_synced_storage_entry::<S, T>(key, init.take().unwrap());
+            if cx.generation() == 0 {
+                // The first generation is rendered on the server side and so must be hydrated.
+                cx.needs_update();
+            }
+            if cx.generation() == 1 {
+                // The first time the vdom is hydrated, we set the correct value from storage and set up the subscription to storage events.
+                storage_entry
+                    .entry
+                    .set(get_from_storage::<S, T>(key_clone, init.take().unwrap()));
+                storage_entry.save_to_storage_on_change();
+                storage_entry.subscribe_to_storage(cx);
+            }
+            *storage_entry.data()
+        } else {
+            // The client is rendered normally, so we can just use the synced storage entry.
+            let storage_entry = new_synced_storage_entry::<S, T>(key, init.take().unwrap());
+            storage_entry.save_to_storage_on_change();
+            storage_entry.subscribe_to_storage(cx);
+            *storage_entry.data()
+        }
+    };
+    signal
+}
+
+/// A hook that creates a StorageEntry with the latest value from storage or the init value if it doesn't exist.
+pub fn use_storage_entry<S, T>(
```

Comment:
Is this hook and the following one still necessary? It's currently unused and seems superseded by use_storage

Reply:
This hook handles syncronization between multiple storage entries. The other hook doesn't handle multiple windows/multiple tabs syncronizing with eachother

Reply:
true, i guess it could be useful to anyone extending this interface